File: ios/CashuWallet/Core/Protocols/PaymentMethodProtocol.swift
import Foundation

// MARK: - Payment Rail Protocol

/// Protocol defining the interface for different payment rails.
/// Implementations include Lightning (current), Bolt12 (future), and Onchain (future).
///
/// Note: Named "PaymentRail" to avoid conflict with CashuDevKit.PaymentMethod enum.
/// This abstraction allows the wallet to support multiple payment rails
/// while maintaining a consistent interface for the UI layer.
protocol PaymentRail {
    /// Unique identifier for this payment method (e.g., "lightning", "bolt12", "onchain")
    var identifier: String { get }
    
    /// Human-readable display name (e.g., "Lightning", "BOLT12", "On-chain")
    var displayName: String { get }
    
    /// Whether this payment method is currently available
    var isAvailable: Bool { get }
    
    /// Create a payment request (invoice) for receiving funds
    /// - Parameters:
    ///   - amount: The amount to request
    ///   - memo: Optional memo/description for the payment
    /// - Returns: A payment request that can be shared with the sender
    func createPaymentRequest(amount: CurrencyAmount, memo: String?) async throws -> PaymentRequest
    
    /// Pay an existing payment request
    /// - Parameter request: The payment request to pay
    /// - Returns: The result of the payment attempt
    func pay(request: PaymentRequest) async throws -> PaymentResult
    
    /// Check the status of a payment
    /// - Parameter paymentId: The unique identifier of the payment
    /// - Returns: The current status of the payment
    func checkPaymentStatus(paymentId: String) async throws -> PaymentStatus
}

// MARK: - Payment Request

/// A request for payment that can be shared with a sender
struct PaymentRequest {
    /// Unique identifier for this request
    let id: String
    
    /// The payment rail this request is for
    let paymentRail: String
    
    /// The requested amount
    let amount: CurrencyAmount
    
    /// Encoded request string (e.g., bolt11 invoice, bip21 URI)
    let encodedRequest: String
    
    /// Optional memo/description
    let memo: String?
    
    /// Expiration timestamp (nil if no expiration)
    let expiresAt: Date?
    
    /// Whether the request has expired
    var isExpired: Bool {
        guard let expiresAt else { return false }
        return Date() > expiresAt
    }
}

// MARK: - Payment Result

/// Result of a payment attempt
struct PaymentResult {
    /// Whether the payment was successful
    let success: Bool
    
    /// Unique identifier for the payment
    let paymentId: String
    
    /// The amount that was actually paid
    let amount: CurrencyAmount
    
    /// Fee paid for the payment
    let fee: CurrencyAmount
    
    /// Payment preimage (for Lightning payments)
    let preimage: String?
    
    /// Error message if payment failed
    let errorMessage: String?
}

// MARK: - Payment Status

/// Status of a payment
enum PaymentStatus {
    case pending
    case completed(preimage: String?)
    case failed(reason: String)
    case expired
    
    var isPending: Bool {
        if case .pending = self { return true }
        return false
    }
    
    var isCompleted: Bool {
        if case .completed = self { return true }
        return false
    }
}

File: android/app/src/main/java/com/cashu/me/Core/Protocols/PaymentMethodProtocol.kt
package com.cashu.me.Core.Protocols

import com.cashu.me.Models.PaymentMethodKind

interface PaymentRail {
    val kind: PaymentMethodKind
    val isAvailable: Boolean

    val identifier: String
        get() = kind.rawValue

    val displayName: String
        get() = kind.displayName

    val iconName: String
        get() = kind.iconName

    suspend fun createPaymentRequest(amount: CurrencyAmount, memo: String? = null): PaymentRequest
    suspend fun pay(request: PaymentRequest): PaymentResult
    suspend fun checkPaymentStatus(paymentId: String): PaymentStatus
}

interface PaymentMethodSupport {
    fun supportsMint(method: PaymentMethodKind): Boolean
    fun supportsMelt(method: PaymentMethodKind): Boolean
}

data class PaymentRequest(
    val id: String,
    val paymentRail: String,
    val amount: CurrencyAmount,
    val encodedRequest: String,
    val memo: String? = null,
    val expiresAtEpochMillis: Long? = null,
) {
    fun isExpired(nowEpochMillis: Long = System.currentTimeMillis()): Boolean {
        val expiresAt = expiresAtEpochMillis ?: return false
        return nowEpochMillis > expiresAt
    }
}

data class PaymentResult(
    val success: Boolean,
    val paymentId: String,
    val amount: CurrencyAmount,
    val fee: CurrencyAmount,
    val preimage: String? = null,
    val errorMessage: String? = null,
)

sealed class PaymentStatus {
    data object Pending : PaymentStatus()
    data class Completed(val preimage: String? = null) : PaymentStatus()
    data class Failed(val reason: String) : PaymentStatus()
    data object Expired : PaymentStatus()

    val isPending: Boolean
        get() = this is Pending

    val isCompleted: Boolean
        get() = this is Completed
}

val PaymentMethodKind.iconName: String
    get() = when (this) {
        PaymentMethodKind.Bolt11 -> "bolt"
        PaymentMethodKind.Bolt12 -> "bolt12"
        PaymentMethodKind.Onchain -> "bitcoin"
    }

val PaymentMethodKind.capabilityLabel: String
    get() = when (this) {
        PaymentMethodKind.Bolt11 -> "Lightning invoice"
        PaymentMethodKind.Bolt12 -> "Reusable offer"
        PaymentMethodKind.Onchain -> "Bitcoin address"
    }
